### ISSUE TITLE
vector3.angleTo with vec on (0,0,0) returns NaN

### ISSUE THREAD

**Issue by felixmariotto** [mrdoob/three.js#17545]
I noticed that vector3.angleTo() returns NaN when either the vector on which it's called or the target vector is at the origin (0,0,0).

```
let vec1 = new THREE.Vector3( -2, 2, 0 );
let vec2 = new THREE.Vector3( 0, 0, 0 );

console.log( vec2.angleTo(vec1) );   // NaN
console.log( vec1.angleTo(vec2) );   // NaN
```

Three.js version : r108
Browser : Chrome
OS : Windows

**Comment by gkjohnson** [comment]
What do you expect it to return?

**Comment by felixmariotto** [comment]
I don't know, either 0 or PI, but NaN broke my code, so I had to debug and to add a conditional in case the angle is NaN.
In most cases, you can't guarantee that the two vectors that you use are not on ( 0, 0, 0 ), so I think it's dangerous to return NaN in this case, I guess most people don't add a conditional every time they use vector3.angleTo()...

**Comment by Mugen87** [comment]
@felixmariotto You are right, this behavior is unfortunate. I suggest the same what Unity does in [Vector3.Angle](https://github.com/Unity-Technologies/UnityCsReference/blob/02d565cf3dd0f6b15069ba976064c75dc2705b08/Runtime/Export/Math/Vector3.cs#L294-L304). `Vector3.angleTo()` would look like so then:
```js
angleTo: function ( v ) {

	var denominator = Math.sqrt( this.lengthSq() * v.lengthSq() );

	if ( denominator === 0 ) return 0;

	var theta = this.dot( v ) / denominator;

	// clamp, to handle numerical problems

	return Math.acos( _Math.clamp( theta, - 1, 1 ) );

}
```

**Comment by WestLangley** [comment]
I am not a fan of returning an incorrect answer. Ever.

three.js does not validate inputs, so there are many methods for which users can pass in problematic values. We can't go down this path, unless we change our policy.

Since the angle to the zero vector does not make sense anyway, I would do nothing.

---

>I suggest the same what Unity does

Unity does not check for zero equality, so that is not exactly what Unity does.

**Comment by Mugen87** [comment]
> Unity does not check for zero equality, so that is not exactly what Unity does.

Oh dear, how meticulous... Well, the approach itself is identical since we do not perform epsilon checks in the math classes in general.

**Comment by WestLangley** [comment]
>Oh dear, how meticulous...

:-)

But the real issue is I don't like returning an incorrect answer. The zero-vector does not define a direction.

**Comment by Mugen87** [comment]
I know your are focused on mathematical correct output. And I do admire this^^. But I also do believe that this change would make the life for devs like @felixmariotto easier. I just want to say that the fewest are aware of such mathematical details and just don't know that `Vector3.angleTo()` could produce a `NaN` value.

**Comment by WestLangley** [comment]
Your proposed solution silently produces an incorrect answer.

**Comment by WestLangley** [comment]
Passing in the zero-vector is a user error. The zero-vector does not define a direction.

The correct solution is to validate inputs, but we do not do that.

Passing in a Vector2 also produces NaN.

**Comment by WestLangley** [comment]
@Mugen87 Ugh... If you want, I guess we can _define_ the angle between a vector and the zero-vector to be zero.

    Return zero if one of the vectors is the zero-vector.

That would be what you are proposing -- with an added comment.

**Comment by gkjohnson** [comment]
I agree with @WestLangley 0 doesn't seem like a reasonable value to return in this case. `AngleTo` implies that the vector can be rotated by the returned value to point in the same direction and that's not possible for a zero vector.

> I also do believe that this change would make the life for devs like @felixmariotto easier

I think silently / untraceably returning that two vectors are pointing in the same direction when they are not could make life harder for other devs, no? Appropriately handling the corner case of trying to get the angle to a zero vector seems like an application-specific concern. Assuming that 0 is a reasonable or appropriate value in all cases doesn't seem right.

> I guess we can define the angle between a vector and the zero-vector to be zero.

Perhaps the docs could just be updated to inform users that the function will return `NaN` if one of the vectors is zero?

**Comment by Mugen87** [comment]
Well...unfortunately we can't ask the developers at Unity but I'm convinced they had a reason for returning `0` in this case. IMO, it's more dev-friendly if `Vector3` handles this specific use case gracefully and avoids returning `NaN`(which breaks all subsequent computations). 

> I guess we can define the angle between a vector and the zero-vector to be zero.

That sounds good to me.

**Comment by WestLangley** [comment]
Doing this
```js
var theta = this.dot( v ) / ( Math.sqrt( this.lengthSq() * v.lengthSq() ) || 1 );
```
will return `pi/2`, instead. By some arguments, that is actually the correct answer.

Also, `projectOnVector( zeroVector )` will return `NaN`.

**Comment by Mugen87** [comment]
> By some arguments, that is actually the correct answer.

Can you please explain a bit?

**Comment by WestLangley** [comment]
Sure. Google "zero vector orthogonal".

But even if we were to agree that the zero vector is orthogonal to every vector, I think the concept of "angle to" requires both vectors to have a non-zero length. So by that argument, passing in the zero vector is a user error -- which is what I was arguing from the beginning. :-)

**Comment by felixmariotto** [comment]
The true issue is not to find the correct answer to this incorrect question, but rather to do something user-friendly. You could stick to mathematical correctness, but still help the users by throwing a console error in this situation, so the debugging would be straightforward.

**Comment by WestLangley** [comment]
@felixmariotto three.js does not validate inputs.

Perhaps the best solution is to state in the docs that both `this` and `v` must not be the zero vector.

---

Why did you pass the zero vector into this method? What was your use case?

**Comment by felixmariotto** [comment]
@WestLangley 

I have this function to get the angle between vec1 and vec2 on vecAngle, in order to do some trigonometry : 

```
function angleThreeVectors( vec1, vecAngle, vec2) {

	vec1.sub( vecAngle );
	vec2.sub( vecAngle );

	angle = vec1.angleTo( vec2 );

	vec1.add( vecAngle );
	vec2.add( vecAngle );

	return angle ;
};
```

But in my application I cannot guarantee that vec1 != vecAngle.
Adding this solved the issue in my case : 
`return angle  || 0`

**Comment by Mugen87** [comment]
From my experience with previous math-related discussions, I know that we reached a dead-end here. The only thing that will be possible is updating the docs. Unless @mrdoob decides for something different...

**Comment by mrdoob** [comment]
`angleTo` doesn't seem to be a hot method (we don't use it in the render loop) so I think it would be okay to throw an error if `denominator` is `0`. In this case it wouldn't be so much about validating inputs, but about avoiding dividing by 0. How does that sound?

**Comment by Mugen87** [comment]
> so I think it would be okay to throw an error if denominator is 0. 

@felixmariotto Would you prefer such a behavior compared to the current one?

**Comment by WestLangley** [comment]
>In this case it wouldn't be so much about validating inputs, but about avoiding dividing by 0. How does that sound?

You can't calculate an angle to a vector that that defines no direction.

This is a user error and an invalid input.

Valid inputs will not divide by zero. If we check for zero here, we are validating input and we will need to do it everywhere.

**Comment by felixmariotto** [comment]
> Would you prefer such a behavior compared to the current one?

Yes as a user I do prefer this behaviour, as it would have help me during the debugging. My input was not always wrong, so it was hard to trace the bug down to the NaN produced by angleTo, without any feedback from Three. I wish I was so good at math as you guys are, I'd not have done the mistake in the first place... But I studied other things.

**Comment by mrdoob** [comment]
@WestLangley 

Out of curiosity... In this case, how do you think @felixmariotto method should look like? Something like this?

```js
function angleThreeVectors( vec1, vecAngle, vec2) {

	if ( vec1.x === 0 && vec1.y === 0 && vec1.z === 0 ) throw new Error( 'vec1 is 0' );
	if ( vec2.x === 0 && vec2.y === 0 && vec2.z === 0 ) throw new Error( 'vec2 is 0' );

	vec1.sub( vecAngle );
	vec2.sub( vecAngle );

	angle = vec1.angleTo( vec2 );

	vec1.add( vecAngle );
	vec2.add( vecAngle );

	return angle ;
};
```

Should we provide a `isZero()` helper?

**Comment by WestLangley** [comment]
@mrdoob The OP is subtracting vectors, so the test needs to go in a different place.

>Should we provide a isZero() helper?

I don't think so.

**Comment by mrdoob** [comment]
> @mrdoob The OP is subtracting vectors, so the test needs to go in a different place.

Oh I see.

Hmm... Don't you see this request similar to the `Error` we throw when a matrix is degenerate?

https://github.com/mrdoob/three.js/blob/dev/src/math/Matrix4.js#L543-L559

**Comment by WestLangley** [comment]
In the matrix case, the user is asking to invert a matrix that is not invertible.

In this case, the user is asking to compute the angle to a vector -- a vector that does not represent a direction.

It's up to you.

**Comment by mrdoob** [comment]
> It's up to you.

I definitely continue thinking that we should not validate inputs. But, in a case by case basis, we can try to make dealing with javascript easier.

In this case, I think it's okay to add a `console.error()` when `denominator` is `0`.

**Comment by saschandroid** [comment]
As the user has to deal with valid inputs anways (test for non-zero vectors), wouldn't it be better to just return  π/2 radians (90°) instead of a console.error? Returning π/2 seems for me to be mathematical correct ... 
Wikipedia:
"In Euclidean space, **two vectors are orthogonal if** and only if their dot product is zero, i.e. they make an angle of 90° (π/2 radians), or **one of the vectors is zero**." (https://en.wikipedia.org/wiki/Orthogonality).
Wikibooks:
"**the angle between the zero vector and any other vector is** defined to be **a right angle**" (https://en.wikibooks.org/wiki/Linear_Algebra/Length_and_Angle_Measures).
Returning (mathematically wrong) NaN (and conole.error) because the user _should_ not try to calculate the angle to/from a zero-vector doesn't look like a good solution for me.

Users would then only have to check for zero-vectors if the returned angle is exactly π/2 and only if this is relevant later on. 

By the way: I use "angleTo" while calculating weighted (angle and surface) vertex normals. That worked perfectly fine because I had to test the angle to be in a certain range anyways. After updating three.js I got thousands of console.errors, so I now have to test every (calculated) vector to be non-zero before using "angleTo".
